src/caches/LRUDiskCache.ts

```diff
@@ -0,0 +1,135 @@
+import { CacheLayer } from './CacheLayer'
+import { LRUDiskCacheOptions, LRUStats } from './typings'
+
+import { outputJSON, readJSON, remove } from 'fs-extra'
+import * as LRU from 'lru-cache'
+import { join } from 'path'
+import * as ReadWriteLock from 'rwlock'
+
+
+export class LRUDiskCache<V> implements CacheLayer<string, V>{
+  
+  private lock: ReadWriteLock
+  private disposed: number
+  private hits = 0
+  private total = 0
+  private lruStorage: LRU.Cache<string, number>
+  private keyToBeDeleted: string
+
+  constructor(private cachePath: string, options: LRUDiskCacheOptions, private readFile=readJSON, private writeFile=outputJSON) {
+    this.hits = 0
+    this.total = 0
+    this.disposed = 0
+    this.keyToBeDeleted = ''
+    this.lock = new ReadWriteLock()
+
+    const dispose = (key: string, timeOfDeath: number): void => {
+      this.keyToBeDeleted = key
+      this.disposed += 1
+    }
+
+    this.lruStorage = new LRU<string, number>({
+      ...options,
+      dispose,
+    })
+
+  }
+
+  public has = (key: string): boolean => this.lruStorage.has(key)
+
+  public getStats = (name='disk-lru-cache'): LRUStats => {
+    const stats = {
+      disposedItems: this.disposed,
+      hitRate: this.total > 0 ? this.hits / this.total : undefined,
+      hits: this.hits,
+      itemCount: this.lruStorage.itemCount,
+      length: this.lruStorage.length,
+      max: this.lruStorage.max,
+      name,
+      total: this.total,
+    }
+    this.hits = 0
+    this.total = 0
+    this.disposed = 0
+    return stats
+  }
+
+  public get = async (key: string): Promise<V | void>  => {
+    const timeOfDeath = this.lruStorage.get(key)
+    if (timeOfDeath === undefined) {
+      this.total += 1
+
+      // if it is an outdated file when stale=false
+      if (this.keyToBeDeleted) {
+        await this.deleteFile(key)
+      }
+      return undefined
+    }
+
+    const pathKey = this.getPathKey(key)
+    try {
+      const data = await new Promise<V>(resolve => {
+        this.lock.readLock(key, async (release: () => void) => {
+          const fileData = await this.readFile(pathKey)
+          release()
+          resolve(fileData)
+        })
+      })
+  
+      this.total += 1
+      this.hits += 1
+  
+      // if it is an outdated file when stale=true
+      if (!this.lruStorage.has(key)) {
```

Comment:
This conditions implies we know exactly how LRU cache works internally. I'd rather explicitly show the file is outdated using the following condition
```suggestion
      if (timeOfDeath < Date.now()) {
```